Importar Librerías

In [87]:
import pandas as pd
import numpy as np
df_train = pd.read_csv("train.csv")
df_test = pd.read_csv("test.csv")

# Modelo Supervisado

* Necesitaremos de un Modelo Supervisado, dado a que nuestra data de entrenamiento nos brinda variables independientes o predictoras así como el resultado de una variable objetivo

In [88]:
df_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


* Contamos con una cantidad de 891 registros y 12 columnas
* Datos faltantes en Edad, Cabina,Embarco

In [89]:
df_train.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

* Cambiamos los datos de sexo a datos numéricos
* Cambiar ambos dataframes

In [90]:
#Cambiamos los Datos de Sexo a tipo numérico... en ambos
df_train["Sex"].replace(("female","male"),(0,1), inplace=True)
df_test["Sex"].replace(("female","male"),(0,1), inplace=True)

In [91]:
df_train["Embarked"].value_counts()

S    644
C    168
Q     77
Name: Embarked, dtype: int64

In [92]:
#Cambiamos los datos de embarque por números
df_train.replace(("Q","S","C"),(0,1,2),inplace=True)
df_test.replace(("Q","S","C"),(0,1,2),inplace=True)

* Resolver el asunto de datos faltantes en la columna "edad"
* Recuerda que al obtener en media no se considera para nada los datos faltantes

In [93]:
#Encontramos la media en cada dataframe (EDAD)
print(df_train["Age"].mean())
df_test["Age"].mean()

29.69911764705882


30.272590361445783

* Tomamos como 30 el valor representativo de la edad

In [94]:
#Reemplazando los datos faltantes con la edad
df_train["Age"] = df_train["Age"].fillna(30)
df_test["Age"] = df_test["Age"].fillna(30)

* Formamos Intervalos para las edades

In [95]:
#Formación de intervalos
bins = [0,8,15,18,25,40,60,100]
names = ["1","2","3","4","5","6","7"]
df_train["Age"] = pd.cut(df_train["Age"],bins,labels=names)
df_test["Age"] = pd.cut(df_test["Age"],bins,labels=names)

* Eliminamos la columna Cabin por su gran cantidad de datos faltante

In [96]:
#Eliminamos la Cabina
df_train.drop(["Cabin"],axis=1,inplace=True)
df_test.drop(["Cabin"],axis=1,inplace=True)

* Eliminamos los datos que nos parezcan irrelevantes en df_train

In [97]:
df_train = df_train.drop(["PassengerId","Name","Ticket"],axis=1)
df_test = df_test.drop(["Name","Ticket"], axis=1)

In [99]:
df_train.dropna(inplace=True)

In [110]:
df_test["Fare"] = df_test["Fare"].fillna(15)

# Modelos de Machine Learning

In [112]:
#Modelos de Machine Learning
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC 
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier 

In [113]:
df_train.tail(3)

Survived  Pclass  Sex Age  SibSp  Parch   Fare  Embarked
888         0       3    0   5      1      2  23.45       1.0
889         1       1    1   5      0      0  30.00       2.0
890         0       3    1   5      0      0   7.75       0.0

In [114]:
#Formamos los apartados de entrenamiento
X_train = np.array(df_train.drop(["Survived"],axis=1))
y_train = np.array(df_train["Survived"])

In [115]:
df_test.head()

PassengerId  Pclass  Sex Age  SibSp  Parch     Fare  Embarked
0          892       3    1   5      0      0   7.8292         0
1          893       3    0   6      1      0   7.0000         1
2          894       2    1   7      0      0   9.6875         0
3          895       3    1   5      0      0   8.6625         1
4          896       3    0   4      1      1  12.2875         1

In [116]:
#Formamos los datos de testeo
X_test = np.array(df_test.drop(["PassengerId"],axis=1))

# Regresión Logística

In [122]:
#Definimos el modelo
algoritmo = LogisticRegression()
#Entrenamos al modelo
algoritmo.fit(X_train,y_train)
#Obtenemos la predicción
y_pred = algoritmo.predict(X_test)
#Obtenemos el Score de entrenamiento
algoritmo.score(X_train,y_train)

0.8031496062992126

# Soporte de Vectores

In [124]:
algoritmo = SVC()
algoritmo.fit(X_train,y_train)
y_pred = algoritmo.predict(X_test)
algoritmo.score(X_train,y_train)

0.6771653543307087

# Vecinos Más cercanos

In [135]:
algoritmo = KNeighborsClassifier(n_neighbors=3)
algoritmo.fit(X_train,y_train)
y_pred = algoritmo.predict(X_test)
algoritmo.score(X_train,y_train)

0.8582677165354331

In [125]:
#Obtener CSV
df = pd.DataFrame(dict(PassengerId = df_test["PassengerId"],Survived = y_pred))
df.to_csv("02_03.csv",index=False)
